server/src/main/java/org/elasticsearch/action/support/ParameterizedTransportAction.java

```diff
@@ -0,0 +1,190 @@
+
+        ClusterService clusterService();
+    }
+
+    @FunctionalInterface
+    private interface Invoker<Req extends ActionRequest, Resp extends ActionResponse> {
+        void invoke(Task task, Req request, ActionListener<Resp> response);
+    }
+
+    @FunctionalInterface
+    private interface ParameterSupplier<Req extends ActionRequest, Resp extends ActionResponse> {
+        Object resolveParameter(Task task, Req request, ActionListener<Resp> channel);
+    }
+
+    private final Invoker<Request, Response> handler;
+
+    @SuppressWarnings("this-escape")
+    protected ParameterizedTransportAction(
+        String actionName,
+        boolean canTripCircuitBreaker,
+        Services services,
+        ActionFilters actionFilters,
+        Writeable.Reader<Request> requestReader,
+        Executor executor
+    ) {
+        super(actionName, canTripCircuitBreaker, services.transportService(), actionFilters, requestReader, executor);
+        handler = resolveHandler(services);
+    }
+
+    @Override
+    protected final void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        handler.invoke(task, request, listener);
+    }
+
+    private Invoker<Request, Response> resolveHandler(Services services) {
+        final List<Method> methods = Arrays.stream(this.getClass().getMethods())
+            .filter(method -> method.isAnnotationPresent(ActionHandler.class))
+            .toList();
+        final Method executeMethod = switch (methods.size()) {
+            case 1 -> methods.getFirst();
+            case 0 -> throw new IllegalStateException(
+                "cannot find @" + ActionHandler.class.getSimpleName() + " annotated method in [" + getClass() + "]"
+            );
+            default -> throw new IllegalStateException(
+                "found multiple ["
+                    + methods.size()
+                    + "] @"
+                    + ActionHandler.class.getSimpleName()
+                    + " annotated methods in ["
+                    + getClass()
+                    + "] : "
+                    + methods
+            );
+        };
+
+        final Class<?>[] parameterTypes = executeMethod.getParameterTypes();
+        final List<ParameterSupplier<Request, Response>> parameterSuppliers = new ArrayList<>(parameterTypes.length);
+
+        for (int i = 0; i < parameterTypes.length; i++) {
+            final Class<?> type = parameterTypes[i];
+            try {
+                ParameterSupplier<Request, Response> param = resolveParameter(type, services);
+                parameterSuppliers.add(param);
+            } catch (IllegalArgumentException e) {
+                throw new ElasticsearchException(
+                    "cannot provide parameter of type [{}] for parameter #{} to {}",
+                    e,
+                    type,
+                    i + 1,
+                    executeMethod
+                );
+            }
+        }
```

Comment:
Most of the work is done here at registration time where the performance impacts are less severe, and we can fail fast if the method has unsupported parameters (or duplicate annotations).